### PR TITLE
Extend elements.json generation with jsdoc and source locations

### PR DESCRIPTION
**Pull request by rictic**
- extract description for element and properties from jsdoc
  - convert constructor names into type names for primitive types
    (e.g. in a polymer properties block `type: String` means the
    `string` type, not `String`
  - thread line and column for source location through to elements.json
  - properly correct source location for inner documents

### CODE REVIEW

File: src/analyzer.ts [hunk 1]
```diff
@@ -104,10 +104,13 @@ export class Analyzer {
...
    */
-  private async _analyzeSource(type: string, contents: string, url: string):
-      Promise<DocumentDescriptor> {
+  private async _analyzeSource(
+      type: string, contents: string, url: string,
+      locationOffset?: LocationOffset): Promise<DocumentDescriptor> {
     let document = this.parse(type, contents, url);
-    return this._analyzeDocument(document);
+    const documentDescriptor = await this._analyzeDocument(document);
+    documentDescriptor.locationOffset = locationOffset;
```
Comment (justinfagnani) [inline src/analyzer.ts]: what about passing this as an optional parameter to `_analyzeDocument()`?

File: src/generate-elements.ts [hunk 2]
```diff
@@ -86,14 +90,18 @@ function serializeElementDescriptor(
...
     events: propChangeEvents,
     metadata: {},
+    sourceLocation:
+        correctSourceLocation(elementDescriptor.sourceLocation, locationOffset)
   };
 }
 
-function serializePropertyDescriptor(p: PropertyDescriptor): Property {
+function serializePropertyDescriptor(
+    p: PropertyDescriptor, locationOffset?: LocationOffset): Property {
```
Comment (justinfagnani) [inline src/generate-elements.ts]: naming: `p` -> `property`?

File: src/generate-elements.ts [hunk 3]
```diff
@@ -266,3 +292,18 @@ function propertyToAttributeName(propertyName: string): string|null {
...
       /([A-Z])/g, (_: string, c1: string) => `-${c1.toLowerCase()}`);
 }
+
+function correctSourceLocation(
+    sourceLocation: SourceLocation,
+    locationOffset?: LocationOffset): SourceLocation|undefined {
+  if (!sourceLocation) {
+    return undefined;
+  }
+  locationOffset = locationOffset || {line: 0, col: 0};
```
Comment (justinfagnani) [inline src/generate-elements.ts]: How about:

```javascript
if (!locationOffset) return sourceLocation;
return sourceLocation && {
  line: ...,
  column: ...,
};
```

File: src/generate-elements.ts [hunk 4]
```diff
@@ -169,7 +180,8 @@ class ElementGatherer implements AnalysisVisitor {
       }
     } else {
       this._elementPaths.set(element, pathToElement);
-      const elem = serializeElementDescriptor(element, pathToElement);
+      const elem =
```
Comment (justinfagnani) [inline src/generate-elements.ts]: having both 'element' and 'elem' identifiers in the same scope could be confusing? How are they different? Can that be in the name?

File: src/html/html-script-finder.ts [hunk 5]
```diff
@@ -44,13 +45,34 @@ export class HtmlScriptFinder implements HtmlEntityFinder {
           entities.push(
               new ImportDescriptor<ASTNode>('html-script', importUrl, node));
         } else {
+          let locationOffset: LocationOffset|undefined = undefined;
```
Comment (justinfagnani) [inline src/html/html-script-finder.ts]: This function is getting long, IMO. Can you pull this logic out?

Comment (justinfagnani) [inline src/html/html-script-finder.ts]: `= undefined` isn't necessary, it's the default value.

File: src/html/html-script-finder.ts [hunk 6]
```diff
@@ -44,13 +45,34 @@ export class HtmlScriptFinder implements HtmlEntityFinder {
...
         }
       }
     });
 
     return entities;
   }
 }
+
+function isLocationInfo(loc: LocationInfo|
+                        ElementLocationInfo): loc is LocationInfo {
```
Comment (justinfagnani) [inline src/html/html-script-finder.ts]: weird alignment... clang does this?

**Comment by justinfagnani** [comment]
It looks like most changes are covered by high-level tests, but I would love to see some more coverage by simple unit tests for things like correctSourceLocation, objectKeyToString, the locationOffset in HtmlScriptFinder, etc. They're border-line boring tests, but should be simple and easier to reason about and fix than looking at offsets from a test file.

File: src/analyzer.ts (hunk 1 above)
Comment (rictic) [inline src/analyzer.ts]: Done

File: src/generate-elements.ts (hunk 2 above)
Comment (rictic) [inline src/generate-elements.ts]: Done

File: src/generate-elements.ts (hunk 3 above)
Comment (rictic) [inline src/generate-elements.ts]: Nice, done.

File: src/generate-elements.ts (hunk 4 above)
Comment (rictic) [inline src/generate-elements.ts]: Good point. Expanded `element` to `elementDescriptor` and `elem` to `element` here to clarify that ambiguity.

File: src/html/html-script-finder.ts (hunk 5 above)
Comment (rictic) [inline src/html/html-script-finder.ts]: Extracted as `getLocationOffsetOfStartOfTextContent`, and removed unnecessary assignment.

File: src/html/html-script-finder.ts (hunk 6 above)
Comment (rictic) [inline src/html/html-script-finder.ts]: Agreed that it's weird, but yep.

**Comment by rictic** [comment]
All comments addressed, PTAL

File: src/html/html-script-finder.ts [hunk 7]
```diff
@@ -75,4 +60,24 @@ export class HtmlScriptFinder implements HtmlEntityFinder {
 function isLocationInfo(loc: LocationInfo|
                         ElementLocationInfo): loc is LocationInfo {
   return 'line' in loc;
+}
+
+function getLocationOffsetOfStartOfTextContent(node: ASTNode) {
```
Comment (justinfagnani) [inline src/html/html-script-finder.ts]: you know... this would be a decent dom5 addition

**Comment by justinfagnani** [comment]
LGTM

**Pull request merged**